### ISSUE TITLE
no-unused-expression rule marks test cases as error

### ISSUE THREAD

**Issue by JimiC** [palantir/tslint#2485]
### Bug Report

- __TSLint version__: 5.0.0
- __TypeScript version__: 2.2.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
it('but there are no changes in package.json the callback will not be called', function () {
        const spy = sinon.spy();
        manageAutoApplyCustomizations(true, userConfig, spy);
        expect(spy.called).to.not.be.true;
      });
```

with `tslint.json` configuration:

```json
{
  "extends": [
    "tslint:latest"
  ],
  "rules": {
    "quotemark": [
      "single"
    ],
    "eofline": true,
    "variable-name": [
      true,
      "allow-pascal-case",
      "allow-leading-underscore"
    ],
    "space-before-function-paren": [
      "error",
      {
        "anonymous": "always",
        "named": "never",
        "asyncArrow": "ignore"
      }
    ],
    "object-literal-sort-keys": false,
    "no-string-literal": false,
    "ordered-imports": [
      false
    ],
    "arrow-parens": false
  },
  "jsRules": {
    "quotemark": [
      "single"
    ],
    "object-literal-sort-keys": false
  }
}
```

#### Actual behavior
`expect(spy.called).to.not.be.true;` violates `no-unused-expression` rule.
#### Expected behavior
`expect(spy.called).to.not.be.true;` should not be mark as a vialation.

**Comment by BobBuehler** [comment]
foo(a).b looks like it **should** violate the rule because the reference of b looks like it has no effect. In truth that is a getter with side effects which tslint doesn't know. I'm not sure there is a way for tslint to be made aware.

This seems to be an undocumented change in v5.0.0. Maybe in the past tslint saw the foo invocation and stopped checking the line.

I've handled this by disabling the rule in my test files.

**Comment by ajafff** [comment]
I have rewritten this rule for v5.x and I'm pretty sure it worked the same
before.

tslint can not know that this is in fact a getter. Even with typescript's
type checker it's not possible.
The only way to handle such expressions would be to add knowledge about
specific libraries (in this case chai) to tslint. That's very unlikely to
happen. eslint also makes no exception for chai.

You can either disable the rule for the whole file or specific lines or
just use https://www.npmjs.com/package/dirty-chai

Am 28.04.2017 20:41 schrieb "Bob Buehler" <notifications@github.com>:

foo(a).b looks like it *should* violate the rule because the reference of b
looks like it has no effect. In truth that is a getter with side effects
which tslint doesn't know. I'm not sure there is a way for tslint to be
made aware.

This seems to be an undocumented change in v5.0.0. Maybe in the past tslint
saw the foo invocation and stopped checking the line.

I've handled this by disabling the rule in my test files.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/palantir/tslint/issues/2485#issuecomment-298075833>,
or mute
the thread

**Comment by JimiC** [comment]
@ajafff Thanks for sharing this enlightening pov. Honestly I didn't knew about this bad practice and after reading about it I totally agree that it should be avoided at all costs.

Currently we also decided to disable the rule on the entire test file, something that is indeed a risk.

I'm not sure but I think that 'chai', in the latest versions, has opted in ability to use function instead of property assertion using the same assertion naming. I'll have to check it out.

**Comment by BobBuehler** [comment]
I have verified that this is an undocumented change for tslint 5 and I agree that tslint shouldn't change.

**Comment by JimiC** [comment]
To clarify, `chai` does not have opted-in ability to use function assertion instead of property assertion using the same assertion naming. If you are looking for something like this, [Should](http://shouldjs.github.io/) as of version `7` has switched to that. Types for `should` are available [here](https://www.npmjs.com/package/@types/should).
